### PR TITLE
Add support for boolean inputs to FusedOp

### LINKED ISSUE #16723: [Bug] fused_op does not support boolean type
@ptrendx I find that the FusedOp does not support the boolean type. The following script will trigger the error.

```python
import mxnet as mx
import numpy as np
from mxnet.gluon import HybridBlock
mx.npx.set_np()

class Foo(HybridBlock):
    def __init__(self, prefix=None, params=None):
        super(Foo, self).__init__(prefix=prefix, params=params)

    def hybrid_forward(self, F, valid_length):
        mask = (F.np.ones((10,)) < valid_length).astype(np.float32)
        mask2 = (F.np.ones((10,)) < valid_length).astype(np.float32)
        mask = mask * F.np.expand_dims(mask2, axis=-1)
        return mask

foo = Foo()
foo.hybridize()
out = foo(mx.np.ones((10,), ctx=mx.gpu()))
print(out)
```

Stack Trace:
```
MXNetError: [02:32:00] src/operator/fusion/fused_op.cu:76: Unknown type enum 7
Stack trace:
  [bt] (0) /home/ubuntu/mxnet/python/mxnet/../../lib/libmxnet.so(dmlc::LogMessageFatal::~LogMessageFatal()+0x32) [0x7f310563bed2]
  [bt] (1) /home/ubuntu/mxnet/python/mxnet/../../lib/libmxnet.so(mxnet::FusedOp::CheckShapesAndTypes(std::vector<mxnet::TBlob, std::allocator<mxnet::TBlob> > const&, std::vector<mxnet::TBlob, std::allocator<mxnet::TBlob> > const&, std::vector<int, std::allocator<int> >*, std::vector<int, std::allocator<int> >*, std::vector<int, std::allocator<int> >*, std::vector<int, std::allocator<int> >*, int*)+0x17b3) [0x7f310b4743c3]
  [bt] (2) /home/ubuntu/mxnet/python/mxnet/../../lib/libmxnet.so(void mxnet::FusedOp::Forward<mshadow::gpu>(nnvm::NodeAttrs const&, mxnet::OpContext const&, std::vector<mxnet::TBlob, std::allocator<mxnet::TBlob> > const&, std::vector<mxnet::OpReqType, std::allocator<mxnet::OpReqType> > const&, std::vector<mxnet::TBlob, std::allocator<mxnet::TBlob> > const&)+0x1a0) [0x7f310b47df50]
  [bt] (3) /home/ubuntu/mxnet/python/mxnet/../../lib/libmxnet.so(mxnet::imperative::PushFCompute(std::function<void (nnvm::NodeAttrs const&, mxnet::OpContext const&, std::vector<mxnet::TBlob, std::allocator<mxnet::TBlob> > const&, std::vector<mxnet::OpReqType, std::allocator<mxnet::OpReqType> > const&, std::vector<mxnet::TBlob, std::allocator<mxnet::TBlob> > const&)> const&, nnvm::Op const*, nnvm::NodeAttrs const&, mxnet::Context const&, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var*> > const&, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var*> > const&, std::vector<mxnet::Resource, std::allocator<mxnet::Resource> > const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, std::vector<unsigned int, std::allocator<unsigned int> > const&, std::vector<mxnet::OpReqType, std::allocator<mxnet::OpReqType> > const&)::{lambda(mxnet::RunContext)#1}::operator()(mxnet::RunContext) const+0x1423) [0x7f3108a01733]
  [bt] (4) /home/ubuntu/mxnet/python/mxnet/../../lib/libmxnet.so(std::_Function_handler<void (mxnet::RunContext), mxnet::imperative::PushFCompute(std::function<void (nnvm::NodeAttrs const&, mxnet::OpContext const&, std::vector<mxnet::TBlob, std::allocator<mxnet::TBlob> > const&, std::vector<mxnet::OpReqType, std::allocator<mxnet::OpReqType> > const&, std::vector<mxnet::TBlob, std::allocator<mxnet::TBlob> > const&)> const&, nnvm::Op const*, nnvm::NodeAttrs const&, mxnet::Context const&, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var*> > const&, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var*> > const&, std::vector<mxnet::Resource, std::allocator<mxnet::Resource> > const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, std::vector<unsigned int, std::allocator<unsigned int> > const&, std::vector<mxnet::OpReqType, std::allocator<mxnet::OpReqType> > const&)::{lambda(mxnet::RunContext)#1}>::_M_invoke(std::_Any_data const&, mxnet::RunContext&&)+0x17) [0x7f3108a01c17]
  [bt] (5) /home/ubuntu/mxnet/python/mxnet/../../lib/libmxnet.so(std::_Function_handler<void (mxnet::RunContext, mxnet::engine::CallbackOnComplete), mxnet::engine::ThreadedEngine::BulkFlush()::{lambda(mxnet::RunContext, mxnet::engine::CallbackOnComplete)#1}>::_M_invoke(std::_Any_data const&, mxnet::RunContext&&, mxnet::engine::CallbackOnComplete&&)+0x1bf) [0x7f310916266f]
  [bt] (6) /home/ubuntu/mxnet/python/mxnet/../../lib/libmxnet.so(mxnet::engine::ThreadedEngine::ExecuteOprBlock(mxnet::RunContext, mxnet::engine::OprBlock*)+0x995) [0x7f3109166475]
  [bt] (7) /home/ubuntu/mxnet/python/mxnet/../../lib/libmxnet.so(void mxnet::engine::ThreadedEnginePerDevice::GPUWorker<(dmlc::ConcurrentQueueType)0>(mxnet::Context, bool, mxnet::engine::ThreadedEnginePerDevice::ThreadWorkerBlock<(dmlc::ConcurrentQueueType)0>*, std::shared_ptr<dmlc::ManualEvent> const&)+0x11d) [0x7f310917ed7d]
  [bt] (8) /home/ubuntu/mxnet/python/mxnet/../../lib/libmxnet.so(std::_Function_handler<void (std::shared_ptr<dmlc::ManualEvent>), mxnet::engine::ThreadedEnginePerDevice::PushToExecute(mxnet::engine::OprBlock*, bool)::{lambda()#4}::operator()() const::{lambda(std::shared_ptr<dmlc::ManualEvent>)#1}>::_M_invoke(std::_Any_data const&, std::shared_ptr<dmlc::ManualEvent>&&)+0x4e) [0x7f310917f02e]
```

We can also manually disable the Fuse OP, which will generate the correct answer.

```python
import mxnet as mx
import numpy as np
import os
from mxnet.gluon import HybridBlock
mx.npx.set_np()

os.environ['MXNET_USE_FUSION'] = '0'

class Foo(HybridBlock):
    def __init__(self, prefix=None, params=None):
        super(Foo, self).__init__(prefix=prefix, params=params)

    def hybrid_forward(self, F, valid_length):
        mask = (F.np.ones((10,)) < valid_length).astype(np.float32)
        mask2 = (F.np.ones((10,)) < valid_length).astype(np.float32)
        mask = mask * F.np.expand_dims(mask2, axis=-1)
        return mask

foo = Foo()
foo.hybridize()
out = foo(mx.np.ones((10,), ctx=mx.gpu()))
print(out)
```

### PR DESCRIPTION
**Pull request by ptrendx**
## Description ##
Fixes #16723 

This PR fixes 2 problems:
 - FusedOp did not support boolean input type as its inputs.
 - During testing the fix for the first problem another problem was spotted (which I believe was previously fixed during the process of working on the fusion PR, but I guess some merge undid it) - the FusedOp did not recompile the code when there was a change of inputs/outputs number of dimensions, which is wrong.

This PR fixes both of those problems and introduces tests for them. 
@sxjscience Please validate the fix.

### CODE REVIEW

File: tests/python/gpu/test_fusion.py [hunk 1]
```diff
@@ -238,6 +238,49 @@ def test_fusion_compiler_cache():
     if num_gpus > 1:
         check_fused_symbol(a+b, ctx=mx.gpu(1), a=arr1, b=arr2)
 
+@with_seed()
+def test_fusion_boolean_inputs():
+    from mxnet.gluon import HybridBlock
+    mx.npx.set_np()
```
Comment (sxjscience) [inline tests/python/gpu/test_fusion.py]: You may use the `@use_np` decorator like in https://github.com/apache/incubator-mxnet/blob/2c02bfffb2bdb4e8f1e5159e08e3250482e5f603/tests/python/unittest/test_numpy_ndarray.py#L28,
https://github.com/apache/incubator-mxnet/blob/2c02bfffb2bdb4e8f1e5159e08e3250482e5f603/tests/python/unittest/test_numpy_ndarray.py#L36-L40

File: tests/python/gpu/test_fusion.py [hunk 2]
```diff
@@ -256,7 +256,7 @@ def hybrid_forward(self, F, valid_length):
     foo = Foo()
     foo.hybridize(static_alloc=True)
     out = foo(mx.np.ones((10,), ctx=mx.gpu()))
-    mx.npx.reset_np()
+    mx.nd.waitall()
```
Comment (sxjscience) [inline tests/python/gpu/test_fusion.py]: You may use `mx.npx.waitall()` to just rely on the numpy APIs.

**Comment by ptrendx** [comment]
@marcoabreu @larroy Could you tell me what is the configuration of the unix-gpu test runners? They make TVM error out and I cannot reproduce it locally.

**Comment by ptrendx** [comment]
@sxjscience Could you validate the fix in your original code?

**Comment by sxjscience** [comment]
@ptrendx I'm testing it locally

**Comment by sxjscience** [comment]
@ptrendx Thanks for the fix. I've confirmed that it solves my previous failure.

**Comment by larroy** [comment]
@ptrendx is in the container in ci/ folder. You can use the exact same run by using ci/build.py you have the same instruction in the output of CI run either at the top or the bottom.

**Pull request merged**